### PR TITLE
Allow `include()` and `include_dependency()` files to resolve to different depots

### LINKED ISSUE #52161: `include_dependency` files that are in a different depot from the package are not found during loading
This is a MWE from a regression in an internal package. The end goal in that package is to cache some arbitrary piece of code using Julia's precompile system so that code can be loaded again quickly. 

To show the issue, let's create a precompiled module (here in a single file) that we want to be invalidated based on some other file changing. This file could potentially be in a package depot (where it probably won't be changed) but we might also have that dependency file be in a devved package where it could be changed and we then want to recompile the module:

```julia
# Foo.jl
module Foo

using Example

readmefile = joinpath(pkgdir(Example), "README.md")
@show readmefile
include_dependency(readmefile)

end

```

Now, let's say we do not want to pollute our standard depot with cache files from this module (and for some reason we also want to store this module file somewhere else) so we do:

```julia
new_depot = joinpath(tempdir(), "depot")
mkpath(new_depot)
pushfirst!(DEPOT_PATH, new_depot)
pushfirst!(LOAD_PATH, new_depot)
foofile = joinpath(new_depot, "Foo.jl")
if !isfile(foofile)
    cp("Foo.jl", foofile)
end
using Foo
```

Upon running this  we can see the path to the dependency file we injected:

```julia
julia> using Foo
┌ Info: Precompiling Foo [top-level]
└ @ Base loading.jl:2442
readmefile = "/Users/kristoffercarlsson/.julia/packages/Example/aqsx3/README.md"```
```

However, when looking what is actually stored in the precompile file, the path to the README file is different:

```julia
cachefile = joinpath(DEPOT_PATH[1], "compiled", "v1.11", "Foo.ji")
io = open(cachefile, "r")
Base.isvalid_cache_header(io)
modules, (includes, _, requires), _ = Base.parse_cache_header(io, cachefile)
includes

julia> includes
2-element Vector{Base.CacheHeaderIncludes}:
 Base.CacheHeaderIncludes( [top-level], "/var/folders/tp/2p4x9ygx48sgsdl1ccg1mp_40000gn/T/depot/Foo.jl", 0x0000000000000084, 0xd614849c, -1.0, String[])
 Base.CacheHeaderIncludes( [top-level], "/var/folders/tp/2p4x9ygx48sgsdl1ccg1mp_40000gn/T/depot/packages/Example/aqsx3/README.md", 0x0000000000000000, 0x00000000, 1.6745681561065354e9, String[])
```

The README.md file is stored with the location `/var/folders/tp/2p4x9ygx48sgsdl1ccg1mp_40000gn/T/depot/packages/Example/aqsx3/README.md"` which is not the same as what we actually used in the `include_dependency` call.

The issue with that is that when we load this module again, it will not find this file and recompile it:

```
julia> using Foo
┌ Debug: Rejecting stale cache file /var/folders/tp/2p4x9ygx48sgsdl1ccg1mp_40000gn/T/depot/compiled/v1.11/Foo.ji because file /var/folders/tp/2p4x9ygx48sgsdl1ccg1mp_40000gn/T/depot/packages/Example/aqsx3/README.md does not exist
└ @ Base loading.jl:3289
Info: Precompiling Foo [top-level
```

My guess to what is happening is that after https://github.com/JuliaLang/julia/pull/49866 the precompile system checks all file that are in a depot path and "reifies" them to be the same as that of the package getting loaded. But this fails in cases like this.

It would be good to have some solution to this because this use case is quite useful.

cc @fatteneder

### PR DESCRIPTION
**Pull request by fatteneder**
Fixes #52161

#52161 is an awkward use of the relocation feature in the sense that it attempts to load the `include()` and `include_dependency` files of a pkg from two separate depots.
The problem there is that the value with which we replace the `@depot` tag for both `include()` and `include_dependency()` files is determined by trying to relocate only the `include()` files. We then end up not finding the `include_dependency()` files.

Solution:
@staticfloat noted that the pkg slugs in depot paths like `@depot/packages/Foo/1a2b3c/src/Foo.jl` are already enough to (weakly) content-address a depot.
This means that we should be able to load any `include()` file of a pkg from any `depot` that contains a precompile cache, provided the hashes match.
The same logic can be extended to `include_dependency()` files, which this PR does.

Note that we continue to use only one file from the `include()` files to determine the depot which we use to relocate all `include()` files. [this behavior is kept from master]
But for `include_dependency()` files we allow each file to resolve to a different depot. This way the MWE given in #52161 should be extendable to two README files being located in two different pkgs that lie in two different depots.

---

Side note: #49866 started with explicitly verifying that all `include()` files come from the same depot. In #52346 this was already relaxed to pick the first depot for which any `include()` file can be resolved to. This works, because if any other file might be missing from that depot then this is caught in `stalecache()`.

### CODE REVIEW

**Review by staticfloat — APPROVED** [review]
LGTM!  Also, can confirm this fixes the constant precompilation issue referenced in https://github.com/JuliaLang/julia/issues/52161#issuecomment-1875738305

**Comment by staticfloat** [comment]
Looks like some tests might need some adjustment though: https://buildkite.com/julialang/julia-master/builds/31836#018cd6aa-caa1-4e75-aa48-3eea5e857605/802-1575

**Comment by staticfloat** [comment]
x86_64-apple-darwin failure should be fixed by https://github.com/JuliaLang/julia/pull/52782, but I don't understand why it only appears on that platform.

**Comment by staticfloat** [comment]
Windows failures look real.

**Comment by fatteneder** [comment]
I could not spot an obvious problem with the test.
My theory is that its related to these warnings
```sh
From worker 9:	┌ Warning: RelocationTestPkg1 [854e1adb-5a97-46bf-a391-1cfe05ac726d] is already loaded from a different path:
      From worker 9:	│   loaded:    v0.1.0 "C:\\buildkite-agent\\builds\\win2k22-amdci6-5\\julialang\\julia-master\\julia-edb66b8b14\\share\\julia\\test\\RelocationTestPkg1\\src\\RelocationTestPkg1.jl"
```
so I updated the test to work with temporarily generated example pkgs to avoid that.

File: test/relocatedepot.jl [hunk 1]
```diff
@@ -93,27 +95,94 @@ if !test_relocated_depot
...
+                        # precompile Foo in depot3
+                        open(joinpath(depot3, "Foo.jl"), write=true) do io
+                            println(io, """
+                            module Foo
+                            using Example1
+                            using Example2
+                            srcfile1 = joinpath(pkgdir(Example1), "src", "Example1.jl")
+                            srcfile2 = joinpath(pkgdir(Example2), "src", "Example2.jl")
+                            @show srcfile1
+                            @show srcfile2
```
Comment (giordano) [inline test/relocatedepot.jl:156]: Are these `@show`'s necessary?  They're spamming all tests.

Comment (fatteneder) [reply test/relocatedepot.jl:156]: > Are these @show's necessary?

No.

**Pull request merged**